src/gc-pages.c

```diff
@@ -96,113 +84,184 @@ static void jl_gc_alloc_region(region_t *region)
             jl_throw(jl_memory_exception);
         }
     }
-    const size_t pages_sz = sizeof(jl_gc_page_t) * pg_cnt;
-    const size_t allocmap_sz = sizeof(uint32_t) * pg_cnt / 32;
-    region->pages = (jl_gc_page_t*)mem;
-    region->allocmap = (uint32_t*)(mem + pages_sz);
-    region->meta = (jl_gc_pagemeta_t*)(mem + pages_sz +allocmap_sz);
-    region->lb = 0;
-    region->ub = 0;
-    region->pg_cnt = pg_cnt;
-#ifdef _OS_WINDOWS_
-    VirtualAlloc(region->allocmap, pg_cnt / 8, MEM_COMMIT, PAGE_READWRITE);
-    VirtualAlloc(region->meta, pg_cnt * sizeof(jl_gc_pagemeta_t),
-                 MEM_COMMIT, PAGE_READWRITE);
-#endif
+
+    jl_gc_pagemeta_t *page_meta = (jl_gc_pagemeta_t*)calloc(pg_cnt, sizeof(jl_gc_pagemeta_t));
+    for (pg = 0; pg < pg_cnt; pg++) {
+        struct jl_gc_metadata_ext info;
+        uint32_t msk;
+        unsigned i;
+        region1_t **pregion1;
+        region0_t **pregion0;
+        jl_gc_pagemeta_t **pmeta;
+
+        char *ptr = mem + (GC_PAGE_SZ * pg);
+        page_meta[pg].data = ptr;
+
+        i = REGION_INDEX(ptr);
+        info.region_i = i % 32;
+        info.region_i32 = i / 32;
+        msk = (1 << info.region_i);
+        if ((memory_map.freemap1[info.region_i32] & msk) == 0)
+            memory_map.freemap1[info.region_i32] |= msk; // has free
+        info.region1 = *(pregion1 = &memory_map.meta1[i]);
+        if (!info.region1)
+            info.region1 = (*pregion1 = (region1_t*)calloc(1, sizeof(region1_t)));
+
+        i = REGION1_INDEX(ptr);
+        info.region1_i = i % 32;
+        info.region1_i32 = i / 32;
+        msk = (1 << info.region1_i);
+        if ((info.region1->freemap0[info.region1_i32] & msk) == 0)
+            info.region1->freemap0[info.region1_i32] |= msk; // has free
+        info.region0 = *(pregion0 = &info.region1->meta0[i]);
+        if (!info.region0)
+            info.region0 = (*pregion0 = (region0_t*)calloc(1, sizeof(region0_t)));
+
+        i = REGION0_INDEX(ptr);
+        info.region0_i = i % 32;
+        info.region0_i32 = i / 32;
+        msk = (1 << info.region0_i);
+        info.region0->freemap[info.region0_i32] |= msk; // is free
+        pmeta = &info.region0->meta[i];
+        info.meta = (*pmeta = &page_meta[pg]);
+    }
```

Comment:
The allocation here can probably be cleaned up later.....  (we can have a version of perm_alloc that guarantees zero filling).